File: generated-code/php/codecraft/TcpReadWrite/Model/EntityProperties.php
<?php

namespace Model {
    require_once 'Model/AttackProperties.php';
    require_once 'Model/BuildProperties.php';
    require_once 'Model/RepairProperties.php';

    /**
     * Entity properties
     */
    class EntityProperties
    {
        /**
         * Size. Entity has a form of a square with side of this length
         */
        public $size;
        /**
         * Score for building this entity
         */
        public $buildScore;
        /**
         * Score for destroying this entity
         */
        public $destroyScore;
        /**
         * Whether this entity can move
         */
        public $canMove;
        /**
         * Number of population points this entity provides, if active
         */
        public $populationProvide;
        /**
         * Number of population points this entity uses
         */
        public $populationUse;
        /**
         * Maximum health points
         */
        public $maxHealth;
        /**
         * Cost to build this first entity of this type. If this is a unit (entity can move), the cost is increased by 1 for each existing unit of this type
         */
        public $initialCost;
        /**
         * If fog of war is enabled, maximum distance at which other entities are considered visible
         */
        public $sightRange;
        /**
         * Amount of resource added to enemy able to collect resource on dealing damage for 1 health point
         */
        public $resourcePerHealth;
        /**
         * Build properties, if entity can build
         */
        public $build;
        /**
         * Attack properties, if entity can attack
         */
        public $attack;
        /**
         * Repair properties, if entity can repair
         */
        public $repair;
    
        function __construct($size, $buildScore, $destroyScore, $canMove, $populationProvide, $populationUse, $maxHealth, $initialCost, $sightRange, $resourcePerHealth, $build, $attack, $repair)
        {
            $this->size = $size;
            $this->buildScore = $buildScore;
            $this->destroyScore = $destroyScore;
            $this->canMove = $canMove;
            $this->populationProvide = $populationProvide;
            $this->populationUse = $populationUse;
            $this->maxHealth = $maxHealth;
            $this->initialCost = $initialCost;
            $this->sightRange = $sightRange;
            $this->resourcePerHealth = $resourcePerHealth;
            $this->build = $build;
            $this->attack = $attack;
            $this->repair = $repair;
        }
    
        /**
         * Read EntityProperties from input stream
         */
        public static function readFrom($stream)
        {
            $size = $stream->readInt32();
            $buildScore = $stream->readInt32();
            $destroyScore = $stream->readInt32();
            $canMove = $stream->readBool();
            $populationProvide = $stream->readInt32();
            $populationUse = $stream->readInt32();
            $maxHealth = $stream->readInt32();
            $initialCost = $stream->readInt32();
            $sightRange = $stream->readInt32();
            $resourcePerHealth = $stream->readInt32();
            if ($stream->readBool()) {
                $build = \Model\BuildProperties::readFrom($stream);
            } else {
                $build = NULL;
            }
            if ($stream->readBool()) {
                $attack = \Model\AttackProperties::readFrom($stream);
            } else {
                $attack = NULL;
            }
            if ($stream->readBool()) {
                $repair = \Model\RepairProperties::readFrom($stream);
            } else {
                $repair = NULL;
            }
            return new EntityProperties($size, $buildScore, $destroyScore, $canMove, $populationProvide, $populationUse, $maxHealth, $initialCost, $sightRange, $resourcePerHealth, $build, $attack, $repair);
        }
        
        /**
         * Write EntityProperties to output stream
         */
        public function writeTo($stream)
        {
            $stream->writeInt32($this->size);
            $stream->writeInt32($this->buildScore);
            $stream->writeInt32($this->destroyScore);
            $stream->writeBool($this->canMove);
            $stream->writeInt32($this->populationProvide);
            $stream->writeInt32($this->populationUse);
            $stream->writeInt32($this->maxHealth);
            $stream->writeInt32($this->initialCost);
            $stream->writeInt32($this->sightRange);
            $stream->writeInt32($this->resourcePerHealth);
            if (is_null($this->build)) {
                $stream->writeBool(false);
            } else {
                $stream->writeBool(true);
                $this->build->writeTo($stream);
            }
            if (is_null($this->attack)) {
                $stream->writeBool(false);
            } else {
                $stream->writeBool(true);
                $this->attack->writeTo($stream);
            }
            if (is_null($this->repair)) {
                $stream->writeBool(false);
            } else {
                $stream->writeBool(true);
                $this->repair->writeTo($stream);
            }
        }
    }
}

File: generated-code/pascal/codecraft/TcpReadWrite/Model/EntityProperties.pas
unit EntityProperties;

interface

uses
    AttackProperties in 'Model/AttackProperties.pas',
    BuildProperties in 'Model/BuildProperties.pas',
    Nullable,
    RepairProperties in 'Model/RepairProperties.pas',
    Stream,
    SysUtils;

type
    // Entity properties
    TEntityProperties = class
        // Size. Entity has a form of a square with side of this length
        size: Int32;
        // Score for building this entity
        buildScore: Int32;
        // Score for destroying this entity
        destroyScore: Int32;
        // Whether this entity can move
        canMove: Boolean;
        // Number of population points this entity provides, if active
        populationProvide: Int32;
        // Number of population points this entity uses
        populationUse: Int32;
        // Maximum health points
        maxHealth: Int32;
        // Cost to build this first entity of this type. If this is a unit (entity can move), the cost is increased by 1 for each existing unit of this type
        initialCost: Int32;
        // If fog of war is enabled, maximum distance at which other entities are considered visible
        sightRange: Int32;
        // Amount of resource added to enemy able to collect resource on dealing damage for 1 health point
        resourcePerHealth: Int32;
        // Build properties, if entity can build
        build: TNullable<TBuildProperties>;
        // Attack properties, if entity can attack
        attack: TNullable<TAttackProperties>;
        // Repair properties, if entity can repair
        repair: TNullable<TRepairProperties>;
        constructor Create(size: Int32; buildScore: Int32; destroyScore: Int32; canMove: Boolean; populationProvide: Int32; populationUse: Int32; maxHealth: Int32; initialCost: Int32; sightRange: Int32; resourcePerHealth: Int32; build: TNullable<TBuildProperties>; attack: TNullable<TAttackProperties>; repair: TNullable<TRepairProperties>);
        // Read EntityProperties from input stream
        class function ReadFrom(stream: TStream): TEntityProperties; static;
        // Write EntityProperties to output stream
        procedure WriteTo(stream: TStream);
        function ToString: ansistring; override;
    end;

implementation

constructor TEntityProperties.Create(size: Int32; buildScore: Int32; destroyScore: Int32; canMove: Boolean; populationProvide: Int32; populationUse: Int32; maxHealth: Int32; initialCost: Int32; sightRange: Int32; resourcePerHealth: Int32; build: TNullable<TBuildProperties>; attack: TNullable<TAttackProperties>; repair: TNullable<TRepairProperties>);
begin
    self.size := size;
    self.buildScore := buildScore;
    self.destroyScore := destroyScore;
    self.canMove := canMove;
    self.populationProvide := populationProvide;
    self.populationUse := populationUse;
    self.maxHealth := maxHealth;
    self.initialCost := initialCost;
    self.sightRange := sightRange;
    self.resourcePerHealth := resourcePerHealth;
    self.build := build;
    self.attack := attack;
    self.repair := repair;
end;

class function TEntityProperties.ReadFrom(stream: TStream): TEntityProperties;
var attack: TNullable<TAttackProperties>;
var attackValue: TAttackProperties;
var build: TNullable<TBuildProperties>;
var buildScore: Int32;
var buildValue: TBuildProperties;
var canMove: Boolean;
var destroyScore: Int32;
var initialCost: Int32;
var maxHealth: Int32;
var populationProvide: Int32;
var populationUse: Int32;
var repair: TNullable<TRepairProperties>;
var repairValue: TRepairProperties;
var resourcePerHealth: Int32;
var sightRange: Int32;
var size: Int32;
begin
    size := stream.ReadInt32;
    buildScore := stream.ReadInt32;
    destroyScore := stream.ReadInt32;
    canMove := stream.ReadBoolean;
    populationProvide := stream.ReadInt32;
    populationUse := stream.ReadInt32;
    maxHealth := stream.ReadInt32;
    initialCost := stream.ReadInt32;
    sightRange := stream.ReadInt32;
    resourcePerHealth := stream.ReadInt32;
    if stream.ReadBoolean then begin
        buildValue := TBuildProperties.ReadFrom(stream);
        build := buildValue;
    end else
        build := nil;
    if stream.ReadBoolean then begin
        attackValue := TAttackProperties.ReadFrom(stream);
        attack := attackValue;
    end else
        attack := nil;
    if stream.ReadBoolean then begin
        repairValue := TRepairProperties.ReadFrom(stream);
        repair := repairValue;
    end else
        repair := nil;
    result := TEntityProperties.Create(size, buildScore, destroyScore, canMove, populationProvide, populationUse, maxHealth, initialCost, sightRange, resourcePerHealth, build, attack, repair);
end;

procedure TEntityProperties.WriteTo(stream: TStream);
var attackValue: TAttackProperties;
var buildValue: TBuildProperties;
var repairValue: TRepairProperties;
begin
    stream.WriteInt32(size);
    stream.WriteInt32(buildScore);
    stream.WriteInt32(destroyScore);
    stream.WriteBoolean(canMove);
    stream.WriteInt32(populationProvide);
    stream.WriteInt32(populationUse);
    stream.WriteInt32(maxHealth);
    stream.WriteInt32(initialCost);
    stream.WriteInt32(sightRange);
    stream.WriteInt32(resourcePerHealth);
    if build.HasValue then begin
        stream.WriteBoolean(true);
        buildValue := build.Value;
        buildValue.WriteTo(stream);
    end else
        stream.WriteBoolean(false);
    if attack.HasValue then begin
        stream.WriteBoolean(true);
        attackValue := attack.Value;
        attackValue.WriteTo(stream);
    end else
        stream.WriteBoolean(false);
    if repair.HasValue then begin
        stream.WriteBoolean(true);
        repairValue := repair.Value;
        repairValue.WriteTo(stream);
    end else
        stream.WriteBoolean(false);
end;

function TEntityProperties.ToString: ansistring;
var attackValue: TAttackProperties;
var buildValue: TBuildProperties;
var repairValue: TRepairProperties;
begin
    result := 'EntityProperties {';
    result += 'size=';
    result += IntToStr(size);
    result += ', ';  
    result += 'buildScore=';
    result += IntToStr(buildScore);
    result += ', ';  
    result += 'destroyScore=';
    result += IntToStr(destroyScore);
    result += ', ';  
    result += 'canMove=';
    result += BoolToStr(canMove);
    result += ', ';  
    result += 'populationProvide=';
    result += IntToStr(populationProvide);
    result += ', ';  
    result += 'populationUse=';
    result += IntToStr(populationUse);
    result += ', ';  
    result += 'maxHealth=';
    result += IntToStr(maxHealth);
    result += ', ';  
    result += 'initialCost=';
    result += IntToStr(initialCost);
    result += ', ';  
    result += 'sightRange=';
    result += IntToStr(sightRange);
    result += ', ';  
    result += 'resourcePerHealth=';
    result += IntToStr(resourcePerHealth);
    result += ', ';  
    result += 'build=';
    if build.HasValue then begin
        buildValue := build.Value;
        result += buildValue.ToString;
    end else
        result += 'nil';
    result += ', ';  
    result += 'attack=';
    if attack.HasValue then begin
        attackValue := attack.Value;
        result += attackValue.ToString;
    end else
        result += 'nil';
    result += ', ';  
    result += 'repair=';
    if repair.HasValue then begin
        repairValue := repair.Value;
        result += repairValue.ToString;
    end else
        result += 'nil';
    result += '}';
end;

end.